### ISSUE TITLE
Excalidraw files are saved as empty files on first save to a Dropbox folder

### ISSUE THREAD

**Issue by kyldvs** [excalidraw/excalidraw#1714]
When saving excalidraw files under a certain file name they come out as 0 bytes (and obviously cannot be loaded). This caused me to lose work in a file before realizing this bug.

This is happening in windows/firefox

I can repro this 100% of the time right now. It always occurs on the first save of the file under a new name. Saving the files a second time with the same name works properly. (This happens even if I previously saved successfully and then save again under a new name with 0 changes)

I'm saving directly to a Dropbox folder, and that appears to be part of the issue. I cannot reproduce when saving to a regular windows directory.

![image](https://user-images.githubusercontent.com/1830497/83835089-bd5c1f80-a6a4-11ea-9a21-e9775987f6e8.png)

**Comment by vjeux** [comment]
@tomayac could you look into this? Thanks!

**Comment by tomayac** [comment]
Since this is Firefox, it's not using the Native File System API, but goes through [`file-save-legacy.mjs`](https://github.com/GoogleChromeLabs/browser-nativefs/blob/master/src/legacy/file-save-legacy.mjs). This implies that all files can *only* be saved to the designated Windows Downloads folder, and that there is *no way* to over-save (instead, you get file names with a growing number at the end). You seem to have sync'ed this Downloads folder to Dropbox.

Adobe Lightroom support [discourages](https://helpx.adobe.com/lightroom-classic/kb/zero-kb-file-exported-lightroom.html) you from saving to a sync'ed folder and there are Dropbox users who [complain](https://www.dropboxforum.com/t5/Dropbox-files-folders/Zero-Byte-files-after-each-save/td-p/16269) about this issue, too. Lightroom provides a reasonable explanation:

*"When exporting, Lightroom's first step is to create a file place holder. If you export to a Dropbox synced folder, this place holder is immediately synced to your cloud storage. When this other process takes over the file, it becomes locked. Lightroom can't add new data to it, so it stays a zero-byte document."*

Most probably Firefox's implementation is to also create such a place holder, before then streaming the to-be-saved data into it. I'm on macOS, but there even files *finished* downloading still carry the "Part" label (which signifies an ongoing download), which supports this place holder theory:

<img width="139" alt="Screen Shot 2020-06-05 at 09 44 53" src="https://user-images.githubusercontent.com/145676/83850432-3729f280-a711-11ea-9883-bf9e656a505d.png">

In conclusion, I'd do what Lightroom suggests. I'm closing this Issue, since the problem is apparently not related to Excalidraw, but feel free to recomment, and I can reopen it.

**Comment by kyldvs** [comment]
> This implies that all files can only be saved to the designated Windows Downloads folder, and that there is no way to over-save (instead, you get file names with a growing number at the end).

I don't understand this part, when testing I picked a non Downloads folder to save the file as, and have free reign to rename it or pick a different folder during the save process. My dropbox and downloads folder are not related at all.

Having to just pick a different folder, is an unfortunate "solution". I save almost everything into dropbox because I have many different computers I work on regularly. I don't understand exactly why Excalidraw is different than other websites that handle this just fine in Firefox (there are many of them!). Is this a particular library Excalidraw is using that has this restriction? If so can it be changed? I don't quite see why it is a requirement for us to follow Lightbox's restrictions unless it's being used somehow?

**Comment by tomayac** [comment]
> I don't understand this part, when testing I picked a non Downloads folder to save the file as, and have free reign to rename it or pick a different folder during the save process. My dropbox and downloads folder are not related at all.

Oh, interesting. Windows seems to give you more freedom then than macOS, where there is no option to change the folder:

<img width="486" alt="Screen Shot 2020-06-05 at 10 25 00" src="https://user-images.githubusercontent.com/145676/83854376-eb7a4780-a716-11ea-935f-c6d014a334e3.png">

> Having to just pick a different folder, is an unfortunate "solution". I save almost everything into dropbox because I have many different computers I work on regularly. I don't understand exactly why Excalidraw is different than other websites that handle this just fine in Firefox (there are many of them!).

Can you try this [demo](https://5509.github.io/demos/blob/) and see if it works? Just click the link and a JSON file should be downloaded. It does essentially the same we do.

> Is this a particular library Excalidraw is using that has this restriction? If so can it be changed? I don't quite see why it is a requirement for us to follow Lightbox's restrictions unless it's being used somehow?

We use [browser-nativefs](https://github.com/GoogleChromeLabs/browser-nativefs). It's absolute standard code:

```js
/**
 * Saves a file to disk using the legacy `<a download>` method.
 * @param {Blob} blob - To-be-saved blob.
 * @param {Object} [options] - Optional options object.
 * @param {string} options.fileName - Suggested file name.
 */
export default async (blob, options = {}) => {
  const a = document.createElement('a');
  a.download = options.fileName || 'Untitled';
  a.href = URL.createObjectURL(blob);
  a.addEventListener('click', () => {
    setTimeout(() => URL.revokeObjectURL(a.href), 30 * 1000);
  });
  a.click();
};
```

**Comment by kyldvs** [comment]
> Can you try this demo and see if it works? Just click the link and a JSON file should be downloaded. It does essentially the same we do.

This works just fine.

![image](https://user-images.githubusercontent.com/1830497/83907250-fb475b00-a719-11ea-979c-f62fa8490434.png)

**Comment by tomayac** [comment]
🤔 It’s probably just small enough.

**Comment by kyldvs** [comment]
Perhaps, but the issue still happens for excalidraw at 788 bytes:

![image](https://user-images.githubusercontent.com/1830497/83907772-d30c2c00-a71a-11ea-828f-29e5d80e2676.png)

**Comment by tomayac** [comment]
Have you brought this to the attention of the Dropbox support team? It really seems like not an issue with Excalidraw.

**Comment by kyldvs** [comment]
I could try, but I feel like they would see it as an Excalidraw issue since it is the only website that I am able to reproduce this issue on. Even your example that is the same method excalidraw uses does not have the issue.

**Comment by slfhstr** [comment]
Saving to a file on Mac/safari also produces zero byte file.
100% reproducible
Happens when saving to a location which is native to the local device (Downloads) and to a sync'ed drive (seafile).
This seems to be multi-browser and a core issue, not browser-dependent
